### PR TITLE
[Dist] save shuffled node/edge ID mapping during graph partition

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

Target on https://github.com/dmlc/dgl/issues/4359.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/partition.py [hunk 1]
```diff
@@ -503,6 +504,10 @@ def partition_graph(g, graph_name, num_parts, out_path, num_hops=1, part_method=
     return_mapping : bool
         If `reshuffle=True`, this indicates to return the mapping between shuffled node/edge IDs
         and the original node/edge IDs.
+    save_orig_nids : bool
+        Indicate whether to save original node IDs as node feature of partitions.
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:508]: For bool type value, use "If True, xxxx" to describe usage, e.g.,
```
If True, save the original node IDs as node features of name ``dgl.ORIG_NID``.
```

File: tools/dispatch_data.py [hunk 2]
```diff
@@ -77,6 +79,8 @@ def main():
     parser.add_argument('--ssh-port', type=int, default=22, help='SSH Port.') 
     parser.add_argument('--process-group-timeout', type=int, default=1800,
                         help='timeout[seconds] for operations executed against the process group')
+    parser.add_argument('--save-orig-nids', action='store_true', help='save original node IDs')
```
Comment (jermainewang) [inline tools/dispatch_data.py:82]: 'Save original node IDs as node features of name dgl.ORIG_NID.'

File: tools/dispatch_data.py [hunk 3]
```diff
@@ -77,6 +79,8 @@ def main():
...
+    parser.add_argument('--save-orig-eids', action='store_true', help='save original edge IDs')
```
Comment (jermainewang) [inline tools/dispatch_data.py:83]: same comment

File: tools/distpartitioning/data_shuffle.py [hunk 4]
```diff
@@ -11,6 +11,8 @@
 import torch
 import torch.distributed as dist
 import torch.multiprocessing as mp
+import logging
+from dgl import backend as F
```
Comment (jermainewang) [inline tools/distpartitioning/data_shuffle.py:15]: No need to import other framework-agnostic backend. The code is PyTorch only.

File: tools/distpartitioning/data_shuffle.py [hunk 5]
```diff
@@ -614,17 +616,32 @@ def gen_dist_partitions(rank, world_size, params):
...
             params.graph_name, params.num_parts, \
             schema_map, rank, node_data, edge_data, num_nodes, num_edges)
     memory_snapshot("CreateDGLObjectsComplete: ", rank)
+
+    # save original node/edge IDs
+    if params.save_orig_nids:
+        for ntype, ntype_id in ntypes_map.items():
+            mask = torch.logical_and(graph_obj.ndata[dgl.NTYPE] == ntype_id,
+                                        graph_obj.ndata['inner_node'])
+            orig_nids = F.boolean_mask(graph_obj.ndata['orig_id'], mask)
```
Comment (jermainewang) [inline tools/distpartitioning/data_shuffle.py:629]: Same question. Can we remove or reuse the old "orig_id"?

File: python/dgl/distributed/partition.py [hunk 6]
```diff
@@ -878,6 +883,19 @@ def get_homogeneous(g, balance_ntypes):
             part.ndata['orig_id'] = F.gather_row(sim_g.ndata[NID], part.ndata['orig_id'])
             part.edata['orig_id'] = F.gather_row(sim_g.edata[EID], part.edata['orig_id'])
 
+        if save_orig_nids:
+            for ntype in g.ntypes:
+                ntype_id = g.get_ntype_id(ntype)
+                inner_node_mask = _get_inner_node_mask(part, ntype_id)
+                node_feats[ntype + '/' + ORIG_NID] = F.boolean_mask(part.ndata['orig_id'],
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/distributed/partition.py:890]: If there are 10 node types, you will save 10 copies of ORIG_NIDs. However, you should only need one.

**Review by jermainewang — CHANGES_REQUESTED** [review]

**Comment by Rhett-Ying** [comment]
As discussed offline, we'd like to change the way of ID storage. I will file a new PR
